go/common/gethencoding/geth_encoding.go

```diff
@@ -226,41 +269,94 @@ func ExtractEthCall(param interface{}) (*gethapi.TransactionArgs, error) {
-	if h.BaseFee != nil {
-		baseFee = h.BaseFee.Uint64()
-	}
+		// calculate the converted hash of the parent, for a correct converted chain
+		// default to the genesis
+		convertedParentHash := common.GethGenesisParentHash
+
+		if h.SequencerOrderNo.Uint64() > common.L2GenesisSeqNo {
+			convertedParentHash, err = enc.storage.FetchConvertedHash(h.ParentHash)
+			if err != nil {
+				enc.logger.Error("Cannot find the converted value for the parent of", log.BatchSeqNoKey, h.SequencerOrderNo)
+				return nil, err
+			}
+		}
 
-	return &types.Header{
-		ParentHash:  h.ParentHash,
-		Root:        h.Root,
-		TxHash:      h.TxHash,
-		ReceiptHash: h.ReceiptHash,
-		Difficulty:  big.NewInt(0),
-		Number:      h.Number,
-		GasLimit:    h.GasLimit,
-		GasUsed:     h.GasUsed,
-		BaseFee:     big.NewInt(0).SetUint64(baseFee),
-		Coinbase:    h.Coinbase,
-		Time:        h.Time,
-		MixDigest:   randomness,
-		Nonce:       types.BlockNonce{},
-	}, nil
+		baseFee := uint64(0)
+		if h.BaseFee != nil {
+			baseFee = h.BaseFee.Uint64()
+		}
+
+		gethHeader := types.Header{
+			ParentHash:      convertedParentHash,
+			UncleHash:       gethcommon.Hash{},
+			Root:            h.Root,
+			TxHash:          h.TxHash,
+			ReceiptHash:     h.ReceiptHash,
+			Difficulty:      big.NewInt(0),
+			Number:          h.Number,
+			GasLimit:        h.GasLimit,
+			GasUsed:         h.GasUsed,
+			BaseFee:         big.NewInt(0).SetUint64(baseFee),
+			Coinbase:        h.Coinbase,
+			Time:            h.Time,
+			MixDigest:       perBatchRandomness,
+			Nonce:           types.BlockNonce{},
+			Extra:           h.SequencerOrderNo.Bytes(),
+			WithdrawalsHash: nil,
+			BlobGasUsed:     nil,
+			ExcessBlobGas:   nil,
+			Bloom:           types.Bloom{},
+		}
+		return &gethHeader, nil
+	})
 }
 
-func CreateEthBlockFromBatch(b *core.Batch) (*types.Block, error) {
-	blockHeader, err := CreateEthHeaderForBatch(b.Header, nil)
+// The Geth "Block" type doesn't expose the header directly.
+// This type is required for adjusting the header.
+type localBlock struct {
+	header       *types.Header
+	uncles       []*types.Header
+	transactions types.Transactions
+	withdrawals  types.Withdrawals
+
+	// caches
+	hash atomic.Value
+	size atomic.Value
+
+	// These fields are used by package eth to track
+	// inter-peer block relay.
+	ReceivedAt   time.Time
+	ReceivedFrom interface{}
+}
+
+func (enc *gethEncodingServiceImpl) CreateEthBlockFromBatch(b *core.Batch) (*types.Block, error) {
+	blockHeader, err := enc.CreateEthHeaderForBatch(b.Header)
 	if err != nil {
 		return nil, fmt.Errorf("unable to create eth block from batch - %w", err)
 	}
 
-	return types.NewBlock(blockHeader, b.Transactions, nil, nil, trie.NewStackTrie(nil)), nil
+	// adjust the header of the returned block to make sure the hashes align
+	lb := localBlock{
+		header:       blockHeader,
+		uncles:       nil,
+		transactions: b.Transactions,
+		withdrawals:  nil,
+	}
+	block := *(*types.Block)(unsafe.Pointer(&lb))
```

Comment:
Maybe worth a comment on this line to say something like "these operations force casting of the localBlock to the geth Block type. This was necessary because we can't access the header directly on geth Block"

which I think is what's going on?

Reply:
I literally have this comment on the localBlock type

Reply:
Oh yeah sorry, I guess maybe worth a `// cast to Block` here but up to you, it slowed down my scanning because it looked like such a scary line lol.